### PR TITLE
`DoFTools`: Extract elasticity modes

### PR DESCRIPTION
**Pull request by peterrum**
translational modes (like in `DoFTools::extract_constant_modes()`) + rotational modes

FYI @richardschu

### CODE REVIEW

File: include/deal.II/dofs/dof_tools.h [hunk 1]
```diff
@@ -1382,6 +1382,30 @@ namespace DoFTools
...
                                std::vector<std::vector<bool>>  &constant_modes);
+
+  /**
+   * Same as the above but additional to the translational modes also
+   * the rotational modes are added, needed to set up AMG for solving
+   * elasticity problems.
+   */
+  template <int dim, int spacedim>
+  void
+  extract_elasticity_modes(const Mapping<dim, spacedim>     &mapping,
```
Comment (luca-heltai) [inline include/deal.II/dofs/dof_tools.h:1393]: The natural way to call these in the mechanical engineering community is actually `rigid_body_modes`. Would you mind renaming this to `rigid_body_modes`? It is actually not only useful for elasticity problems, but for any translation and rotation invariant equation.

**Review by luca-heltai — APPROVED** [review]

File: include/deal.II/dofs/dof_tools.h (hunk 1 above)
Comment (tamiko) [reply include/deal.II/dofs/dof_tools.h:1393]: I agree. Let's call it `rigid_body_modes`.

File: source/dofs/dof_tools.cc [hunk 2]
```diff
@@ -43,6 +43,8 @@
 #include <deal.II/lac/sparsity_pattern_base.h>
 #include <deal.II/lac/vector.h>
 
+#include <deal.II/numerics/vector_tools.h>
```
Comment (drwells) [inline source/dofs/dof_tools.cc:46]:
```suggestion
#include <deal.II/numerics/vector_tools_interpolate.h>
```

File: tests/mpi/extract_constant_modes_05.cc [hunk 3]
```diff
@@ -0,0 +1,87 @@
+// ------------------------------------------------------------------------
```
Comment (tjhei) [inline tests/mpi/extract_constant_modes_05.cc:1]: maybe rename the tests to reflect the function name?

File: source/dofs/dof_tools.cc [hunk 4]
```diff
@@ -1407,6 +1409,130 @@ namespace DoFTools
...
+    extract_elasticity_modes(const Mapping<dim, spacedim>     &mapping,
+                             const DoFHandler<dim, spacedim>  &dof_handler,
+                             const ComponentMask              &component_mask,
+                             const unsigned int                mg_level,
+                             std::vector<std::vector<double>> &elasticity_modes)
+    {
+      AssertDimension(dim, spacedim);
+      AssertDimension(component_mask.n_selected_components(), dim);
+
+      const unsigned int n_modes = RigidBodyMotion<dim>::n_modes;
```
Comment (tjhei) [inline source/dofs/dof_tools.cc:1508]: constexpr?

File: source/lac/trilinos_precondition_ml.cc [hunk 5]
```diff
@@ -128,57 +128,70 @@ namespace TrilinosWrappers
...
-    const size_type constant_modes_dimension = constant_modes.size();
-    ptr_distributed_constant_modes = std::make_unique<Epetra_MultiVector>(
-      domain_map, constant_modes_dimension > 0 ? constant_modes_dimension : 1);
-    Assert(ptr_distributed_constant_modes, ExcNotInitialized());
-    Epetra_MultiVector &distributed_constant_modes =
-      *ptr_distributed_constant_modes;
-
-    if (constant_modes_dimension > 0)
+    const auto run = [&](const auto &constant_modes) {
+      const Epetra_Map &domain_map = matrix.OperatorDomainMap();
```
Comment (tjhei) [inline source/lac/trilinos_precondition_ml.cc:132]: a test for this new code would be great as well

**Review by tjhei — APPROVED** [review]
:+1:

**Comment by peterrum** [comment]
@luca-heltai @drwells @tjhei  Thank you for all the comments. I have applied all of them (except one)!

File: include/deal.II/dofs/dof_tools.h [hunk 6]
```diff
@@ -1382,6 +1382,30 @@ namespace DoFTools
                                const DoFHandler<dim, spacedim> &dof_handler,
                                const ComponentMask             &component_mask,
                                std::vector<std::vector<bool>>  &constant_modes);
+
+  /**
+   * Same as the above but additional to the translational modes also
+   * the rotational modes are added, needed to set up AMG for solving
```
[the remark this answers is not in the document]
Comment (peterrum) [reply include/deal.II/dofs/dof_tools.h]: We already have a function for translational modes. We can add one for rotational modes in the future if someone needs it.

**Pull request merged**